File: perf/benchmarks/apps/word_count/word_count.go
// Word-count — typical text-processing application benchmark (Go).
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func genText(n int64) string {
	seed := int64(12345)
	var sb strings.Builder
	sb.Grow(int(n) * 7)
	for i := int64(0); i < n; i++ {
		seed = (seed*1103515245 + 12345) & 0x7FFFFFFF
		idx := seed % 1024
		fmt.Fprintf(&sb, "w%04d", idx)
		if i%12 == 11 {
			sb.WriteByte('\n')
		} else {
			sb.WriteByte(' ')
		}
	}
	return sb.String()
}

func main() {
	n := int64(1_000_000)
	if len(os.Args) > 1 {
		v, _ := strconv.ParseInt(os.Args[1], 10, 64)
		n = v
	}
	text := genText(n)

	t0 := time.Now()
	counts := make(map[string]int64, 2048)
	i, sz := 0, len(text)
	for i < sz {
		for i < sz && (text[i] == ' ' || text[i] == '\n') {
			i++
		}
		start := i
		for i < sz && text[i] != ' ' && text[i] != '\n' {
			i++
		}
		if i > start {
			counts[text[start:i]]++
		}
	}
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0

	var cs int64
	for k, v := range counts {
		idx, _ := strconv.ParseInt(k[1:], 10, 64)
		cs ^= idx*1000003 + v
	}
	fmt.Printf("CHECKSUM:%d\n", cs)
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/apps/word_count/word_count.rs
// Word-count — typical text-processing application benchmark (Rust).
use std::collections::HashMap;
use std::env;
use std::fmt::Write;
use std::time::Instant;

fn gen_text(n: i64) -> String {
    let mut seed: i64 = 12345;
    let mut s = String::with_capacity((n as usize) * 7);
    for i in 0..n {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7FFFFFFF;
        let idx = seed % 1024;
        let _ = write!(&mut s, "w{:04}", idx);
        if i % 12 == 11 {
            s.push('\n');
        } else {
            s.push(' ');
        }
    }
    s
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: i64 = if args.len() > 1 { args[1].parse().unwrap_or(1_000_000) } else { 1_000_000 };
    let text = gen_text(n);
    let bytes = text.as_bytes();

    let t0 = Instant::now();
    let mut counts: HashMap<&str, i64> = HashMap::with_capacity(2048);
    let mut i: usize = 0;
    let sz = bytes.len();
    while i < sz {
        while i < sz && (bytes[i] == b' ' || bytes[i] == b'\n') { i += 1; }
        let start = i;
        while i < sz && bytes[i] != b' ' && bytes[i] != b'\n' { i += 1; }
        if i > start {
            // Safe: ASCII-only generator above, so [start..i] is valid UTF-8.
            let key = unsafe { std::str::from_utf8_unchecked(&bytes[start..i]) };
            *counts.entry(key).or_insert(0) += 1;
        }
    }
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;

    let mut cs: i64 = 0;
    for (k, v) in counts.iter() {
        let idx: i64 = k[1..].parse().unwrap_or(0);
        cs ^= idx.wrapping_mul(1000003).wrapping_add(*v);
    }
    println!("CHECKSUM:{}", cs);
    println!("ELAPSED_MS:{:.6}", ms);
}
